### PR TITLE
Add functions to Import/Clear/Export the WCIF

### PR DESCRIPTION
**Pull request by viroulep**
I wanted to create a pull request on top of #46, but it wouldn't let me open it in this repository because the base branch would have been in my fork too.
Therefore this PR contains all the code from #46; the diff will shrink when #46 is merged, but at the moment I would recommend reviewing this by taking a look at the last 3 commits individually, or going to [the diff between `fixes` and `import-export-wcif`](https://github.com/viroulep/compscript/compare/fixes...viroulep:compscript:import-export-wcif)!

I started by implementing the `ExportWCIF` function, and ran into an expected behavior from `nodemon` I did not see coming, where emitting a json in a subfolder would trigger a rebuild of the application and reset the connection in the browser.
I worked around this by creating a default `OUTPUT_BASE` folder, which I told `nodemon` to ignore through `-i outputs`.
I have created a `dev-server` target in the `package.json` to be able to conveniently run `npm run dev-server` without having to remember the correct command line every time.

As far as the implementation goes it was quite straightforward, but I did not invest into error checking and the tool will just crash if it can't write the file.

The `ImportWCIF` function is quite straightforward too, and load a file relative to `SCRIPT_BASE`.

I also took the liberty of implementing a `ClearWCIF` function: the name may need to be adjusted, since the purpose is not to reset the whole WCIF but basically provide a "clean" WCIF on top of which the script will run (no assignments, no groups, no unexpected extensions).

It's actually awesome just to make sure anything created by another tool is gone.
If I'm not mistaken there is not such a function even for the WCA website admins; you would have to go into rails' console to clean up the WCIF manually (which I believe I've done several times in the past when some competitions' state got messed up).

### CODE REVIEW

File: functions/wcif.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    ctx.competition.persons.forEach((person) => {
+      // FIXME: should we clean up "roles" too actually?
+      // Changing "delegate"/"organizers"/"trainee-delegate" would have no effect
+      // as they are - expectedly! - not patchable on the WCA, but any "staff-*"
+      // should likely be removed too.
+      person.assignments = []
+    })
+    ctx.competition.schedule.venues.forEach((venue) => {
+      venue.rooms.forEach((room) => {
+        room.extensions = []
```
Comment (timreyn) [inline functions/wcif.js:22]: Maybe this should only clear compscript extensions? e.g. if another program like groupifier sets extensions, this probably shouldn't clear those.

File: functions/wcif.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+const fs = require('fs')
+
+const ClearWCIF = {
+  name: 'ClearWCIF',
+  // FIXME: quick helper to make sure we start "from scratch", maybe also
```
Comment (timreyn) [inline functions/wcif.js:5]: I would also suggest clearing extensions from both persons and competition as well.

File: functions/wcif.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  args: [],
+  outputType: 'Void',
+  usesContext: true,
+  mutations: ['schedule', 'persons'],
+  implementation: (ctx, filename) => {
+    ctx.competition.persons.forEach((person) => {
+      // FIXME: should we clean up "roles" too actually?
+      // Changing "delegate"/"organizers"/"trainee-delegate" would have no effect
+      // as they are - expectedly! - not patchable on the WCA, but any "staff-*"
+      // should likely be removed too.
```
Comment (timreyn) [inline functions/wcif.js:17]: I agree with cleaning up roles -- this also affects staff-dataentry, for example.

File: functions/wcif.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+const ImportWCIF = {
+  name: 'ImportWCIF',
+  docs: 'Import the WCIF from a json file',
+  args: [
+    {
+      name: 'filename',
+      type: 'String',
+      docs: 'WCIF filename (relative to SCRIPT_BASE)',
+      defaultValue: 'input.json',
```
Comment (timreyn) [inline functions/wcif.js:40]: Maybe don't use a defaultValue ? I think it would be better to be explicit.

File: functions/wcif.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+const ImportWCIF = {
+  name: 'ImportWCIF',
+  docs: 'Import the WCIF from a json file',
+  args: [
+    {
+      name: 'filename',
+      type: 'String',
+      docs: 'WCIF filename (relative to SCRIPT_BASE)',
```
Comment (timreyn) [inline functions/wcif.js:39]: Why is ImportWCIF relative to SCRIPT_BASE, and ExportWCIF relative to OUTPUT_BASE?

File: functions/wcif.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      defaultValue: 'input.json',
+    },
+  ],
+  outputType: 'String',
+  usesContext: true,
+  implementation: (ctx, filename) => {
+    var fullPath = `${process.env.SCRIPT_BASE || '.'}/${filename}`
```
Comment (timreyn) [inline functions/wcif.js:46]: What do you think about adding the competitionId in the path? That way you can make sure you don't accidentally import from a different competition.

File: outputs/.keep
Comment (timreyn) [inline outputs/.keep:1]: Is this file intentional? I thought the .gitignore ignores this?

Comment (viroulep) [reply outputs/.keep:1]: Yes: since git cannot add an empty folder to the tree it needed at least some kind of file to make sure `outputs` exists when someone clones the repo.

But at the same time we do want to ignore everything in the directory since it will be user-specific data.

File: functions/wcif.js (hunk 6 above)
Comment (viroulep) [reply functions/wcif.js:46]: I knew I had forgotten something; I had a slightly different but arguably "better" idea: make sure the imported competition has the same id as the competition we're currently working on!
Sounds good?

File: functions/wcif.js (hunk 4 above)
Comment (viroulep) [reply functions/wcif.js:40]: Sounds good to me, I'll update this!

File: functions/wcif.js (hunk 5 above)
Comment (viroulep) [reply functions/wcif.js:39]: Initially my thought process was that it would make sense to have all the data required to (re)create the assignments at the same place.
But maybe we're fine with just having a "WCIF_DATA_FOLDER" directory serving as a base for all import/export operations, and where we let the user organize it how ever they want.
In which case I would drop the commit about always setting up `SCRIPT_BASE`, and I would rename `OUTPUT_BASE` to `WCIF_DATA_FOLDER` (or anything else clearly indicating what we aim to store there :p); what do you think?

File: .env.DEV [hunk 7]
```diff
@@ -9,3 +9,5 @@ USE_CDN=
 API_KEY='example-application-id'
 API_SECRET='example-secret'
 COOKIE_SECRET='b8f6959746b617d9c9e90ce4a4b6e0'
+SCRIPT_BASE=scripts
+OUTPUT_BASE=outputs
```
[the remark this answers is not in the document]
Comment (viroulep) [reply .env.DEV]: Done!

File: README.md [hunk 8]
```diff
@@ -10,7 +10,7 @@ Node must be installed on your machine.
 
 ```
 $ npm install
-$ ENV=DEV npx nodemon .
+$ npm run dev-server
 ```
 
 `ENV=DEV` uses a dev WCA environment running on the same machine. If you would like to use the production WCA site, you need to:
```
[the remark this answers is not in the document]
Comment (viroulep) [reply README.md]: Done :)

File: functions/wcif.js (hunk 5 above)
Comment (viroulep) [reply functions/wcif.js]: Fixed by making the import/export relative to `WCIF_DATA_FOLDER`.

File: functions/wcif.js (hunk 6 above)
Comment (viroulep) [reply functions/wcif.js]: I'm not a big fan of imposing a directory structure in a context where I'd personally just point `SCRIPT_BASE` to a git repo containing the scripts for a specific competition, but I went ahead and implemented both the id check and the inclusion of the competition id in the path (and I also updated the documentation to clarify this).

**Comment by viroulep** [comment]
Thanks for the quick review!

I just pushed some changes, hopefully I did not overlook any of your comments, feel free to close/re-open threads to make sure I did not miss something :)
